File: packages/sdk/src/crypto.js
/**
 * AXIP SDK — Cryptographic Identity
 *
 * ed25519 keypair generation, message signing, and signature verification.
 * Each agent gets a persistent identity stored at ~/.axip/<name>/identity.json.
 * Uses tweetnacl for all cryptographic operations.
 */

import nacl from 'tweetnacl';
import { readFileSync, writeFileSync, mkdirSync, existsSync, chmodSync } from 'fs';
import { homedir } from 'os';
import { join } from 'path';

// ─── Base64 Helpers ────────────────────────────────────────────

export function toBase64(uint8Array) {
  return Buffer.from(uint8Array).toString('base64');
}

export function fromBase64(b64String) {
  return new Uint8Array(Buffer.from(b64String, 'base64'));
}

// ─── Key Operations ────────────────────────────────────────────

/**
 * Generate a new ed25519 keypair.
 * @returns {{ publicKey: Uint8Array, secretKey: Uint8Array }}
 */
export function generateKeypair() {
  return nacl.sign.keyPair();
}

/**
 * Format a public key for the AXIP message envelope.
 * @param {Uint8Array} publicKey
 * @returns {string} "ed25519:<base64>"
 */
export function formatPubkey(publicKey) {
  return `ed25519:${toBase64(publicKey)}`;
}

/**
 * Extract raw public key bytes from a formatted pubkey string.
 * @param {string} formatted - "ed25519:<base64>"
 * @returns {Uint8Array}
 */
export function parsePubkey(formatted) {
  const prefix = 'ed25519:';
  if (!formatted.startsWith(prefix)) {
    throw new Error(`Invalid pubkey format: ${formatted}`);
  }
  return fromBase64(formatted.slice(prefix.length));
}

// ─── Signing and Verification ──────────────────────────────────

/**
 * Sign a message with the agent's secret key.
 * @param {string} message - The message string to sign
 * @param {Uint8Array} secretKey - 64-byte ed25519 secret key
 * @returns {string} "ed25519:<base64 signature>"
 */
export function sign(message, secretKey) {
  const messageBytes = new TextEncoder().encode(message);
  const signature = nacl.sign.detached(messageBytes, secretKey);
  return `ed25519:${toBase64(signature)}`;
}

/**
 * Verify a detached signature.
 * @param {string} message - The original message string
 * @param {string} signature - "ed25519:<base64 signature>"
 * @param {Uint8Array} publicKey - 32-byte ed25519 public key
 * @returns {boolean}
 */
export function verify(message, signature, publicKey) {
  try {
    const prefix = 'ed25519:';
    if (!signature.startsWith(prefix)) return false;
    const sigBytes = fromBase64(signature.slice(prefix.length));
    const messageBytes = new TextEncoder().encode(message);
    return nacl.sign.detached.verify(messageBytes, sigBytes, publicKey);
  } catch {
    return false;
  }
}

// ─── Identity Persistence ──────────────────────────────────────

/**
 * Load an existing identity or create a new one.
 * Identities are stored at ~/.axip/<name>/identity.json with chmod 600.
 *
 * @param {string} name - Agent name (used as directory name)
 * @returns {{ agentId: string, publicKey: Uint8Array, secretKey: Uint8Array, pubkeyFormatted: string }}
 */
export function loadOrCreateIdentity(name = 'default') {
  const axipDir = join(homedir(), '.axip', name);
  const identityPath = join(axipDir, 'identity.json');

  if (existsSync(identityPath)) {
    // Load existing identity
    const data = JSON.parse(readFileSync(identityPath, 'utf-8'));
    const publicKey = fromBase64(data.publicKey);
    const secretKey = fromBase64(data.secretKey);

    return {
      agentId: data.agentId,
      publicKey,
      secretKey,
      pubkeyFormatted: formatPubkey(publicKey)
    };
  }

  // Generate new identity
  const keypair = generateKeypair();

  // Derive a human-readable agent ID from the public key
  const agentId = `${name}-${toBase64(keypair.publicKey).slice(0, 8).replace(/[+/=]/g, 'x')}`;

  const data = {
    agentId,
    publicKey: toBase64(keypair.publicKey),
    secretKey: toBase64(keypair.secretKey),
    createdAt: new Date().toISOString()
  };

  // Write with restricted permissions
  mkdirSync(axipDir, { recursive: true });
  writeFileSync(identityPath, JSON.stringify(data, null, 2));
  chmodSync(identityPath, 0o600);

  console.log(`[crypto] New identity created: ${agentId}`);
  console.log(`[crypto] Stored at: ${identityPath}`);

  return {
    agentId,
    publicKey: keypair.publicKey,
    secretKey: keypair.secretKey,
    pubkeyFormatted: formatPubkey(keypair.publicKey)
  };
}

File: packages/axip-python/src/axip/crypto.py
"""
AXIP SDK — Cryptographic Identity

Ed25519 keypair generation, message signing, and signature verification.
Each agent gets a persistent identity stored at ~/.axip/<name>/identity.json.
Uses PyNaCl for all cryptographic operations (compatible with JS tweetnacl).
"""

import base64
import json
import os
import stat
from pathlib import Path
from typing import NamedTuple

import nacl.signing
import nacl.encoding


# ─── Types ────────────────────────────────────────────────────────


class Identity(NamedTuple):
    agent_id: str
    signing_key: nacl.signing.SigningKey   # 32-byte seed → 64-byte secret
    verify_key: nacl.signing.VerifyKey     # 32-byte public key
    pubkey_formatted: str                  # "ed25519:<base64>"


# ─── Base64 Helpers ───────────────────────────────────────────────


def to_base64(data: bytes) -> str:
    return base64.b64encode(data).decode("ascii")


def from_base64(s: str) -> bytes:
    return base64.b64decode(s)


# ─── Key Format ───────────────────────────────────────────────────


def format_pubkey(verify_key: nacl.signing.VerifyKey) -> str:
    return f"ed25519:{to_base64(bytes(verify_key))}"


def parse_pubkey(formatted: str) -> nacl.signing.VerifyKey:
    prefix = "ed25519:"
    if not formatted.startswith(prefix):
        raise ValueError(f"Invalid pubkey format: {formatted}")
    raw = from_base64(formatted[len(prefix):])
    return nacl.signing.VerifyKey(raw)


# ─── Signing and Verification ─────────────────────────────────────


def sign(message: str, signing_key: nacl.signing.SigningKey) -> str:
    """
    Sign a string message. Returns "ed25519:<base64 signature>".
    Compatible with JS tweetnacl detached signatures.
    """
    msg_bytes = message.encode("utf-8")
    signed = signing_key.sign(msg_bytes)
    # PyNaCl .sign() returns signature + message; extract just signature
    sig_bytes = signed.signature
    return f"ed25519:{to_base64(sig_bytes)}"


def verify(message: str, signature: str, verify_key: nacl.signing.VerifyKey) -> bool:
    """
    Verify a detached Ed25519 signature.
    """
    try:
        prefix = "ed25519:"
        if not signature.startswith(prefix):
            return False
        sig_bytes = from_base64(signature[len(prefix):])
        msg_bytes = message.encode("utf-8")
        verify_key.verify(msg_bytes, sig_bytes)
        return True
    except Exception:
        return False


# ─── Keypair ──────────────────────────────────────────────────────


def generate_keypair() -> tuple[nacl.signing.SigningKey, nacl.signing.VerifyKey]:
    signing_key = nacl.signing.SigningKey.generate()
    return signing_key, signing_key.verify_key


# ─── Identity Persistence ─────────────────────────────────────────


def load_or_create_identity(name: str = "default") -> Identity:
    """
    Load an existing identity or create a new one.
    Stored at ~/.axip/<name>/identity.json with mode 600.
    Format is compatible with the JS SDK.
    """
    axip_dir = Path.home() / ".axip" / name
    identity_path = axip_dir / "identity.json"

    if identity_path.exists():
        data = json.loads(identity_path.read_text())
        # JS SDK stores the full 64-byte tweetnacl secret key (seed + pubkey)
        # PyNaCl SigningKey accepts the 32-byte seed only
        raw_secret = from_base64(data["secretKey"])
        if len(raw_secret) == 64:
            # tweetnacl format: first 32 bytes are the seed
            seed = raw_secret[:32]
        else:
            seed = raw_secret
        signing_key = nacl.signing.SigningKey(seed)
        verify_key = signing_key.verify_key

        return Identity(
            agent_id=data["agentId"],
            signing_key=signing_key,
            verify_key=verify_key,
            pubkey_formatted=format_pubkey(verify_key),
        )

    # Generate new identity
    signing_key, verify_key = generate_keypair()
    pubkey_b64 = to_base64(bytes(verify_key))
    # Build agent_id like the JS SDK: name + first 8 chars of pubkey (sanitized)
    suffix = pubkey_b64[:8].replace("+", "x").replace("/", "x").replace("=", "x")
    agent_id = f"{name}-{suffix}"

    # Store in JS-compatible format:
    # secretKey = seed (32 bytes) + pubkey (32 bytes) concatenated = 64 bytes
    seed_bytes = bytes(signing_key)
    pubkey_bytes = bytes(verify_key)
    secret_64 = seed_bytes + pubkey_bytes

    data = {
        "agentId": agent_id,
        "publicKey": pubkey_b64,
        "secretKey": to_base64(secret_64),
        "createdAt": __import__("datetime").datetime.now(__import__("datetime").timezone.utc).isoformat().replace("+00:00", "Z"),
    }

    axip_dir.mkdir(parents=True, exist_ok=True)
    identity_path.write_text(json.dumps(data, indent=2))
    identity_path.chmod(stat.S_IRUSR | stat.S_IWUSR)  # 600

    print(f"[axip] New identity created: {agent_id}")
    print(f"[axip] Stored at: {identity_path}")

    return Identity(
        agent_id=agent_id,
        signing_key=signing_key,
        verify_key=verify_key,
        pubkey_formatted=format_pubkey(verify_key),
    )
